### PR TITLE
added `BaseBandModulationParams` with `time_on_air_us` to calculate time on air of a packet

### PR DESCRIPTION
**Pull request by ilya-epifanov**

### CODE REVIEW

**Review by lthiery — APPROVED** [review]

File: modulation/src/lib.rs [hunk 1]
```diff
@@ -16,9 +16,9 @@ pub enum Bandwidth {
     _500KHz,
 }
 
-impl From<Bandwidth> for u32 {
-    fn from(value: Bandwidth) -> Self {
-        match value {
+impl Bandwidth {
+    pub const fn hz(self) -> u32 {
```
Comment (ilya-epifanov) [inline modulation/src/lib.rs:20]: new API

File: modulation/src/lib.rs [hunk 2]
```diff
@@ -47,9 +53,9 @@ pub enum SpreadingFactor {
     _12,
 }
 
-impl From<SpreadingFactor> for u32 {
-    fn from(sf: SpreadingFactor) -> Self {
-        match sf {
+impl SpreadingFactor {
+    pub const fn factor(self) -> u32 {
```
Comment (ilya-epifanov) [inline modulation/src/lib.rs:57]: new API

File: modulation/src/lib.rs [hunk 3]
```diff
@@ -72,3 +84,169 @@ pub enum CodingRate {
     _4_7,
     _4_8,
 }
+
+impl CodingRate {
+    pub const fn denom(&self) -> u32 {
```
Comment (ilya-epifanov) [inline modulation/src/lib.rs:89]: new API

File: modulation/src/lib.rs [hunk 4]
```diff
@@ -72,3 +84,169 @@ pub enum CodingRate {
...
+            CodingRate::_4_7 => 7,
+            CodingRate::_4_8 => 8,
+        }
+    }
+}
+
+/// LoRa modulation parameters barring frequency
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct BaseBandModulationParams {
```
Comment (ilya-epifanov) [inline modulation/src/lib.rs:102]: new API

File: modulation/src/lib.rs [hunk 5]
```diff
@@ -72,3 +84,169 @@ pub enum CodingRate {
...
+    /// eagerly pre-calculated symbol duration in microseconds
+    t_sym_us: u32,
+}
+
+impl BaseBandModulationParams {
+    /// Create a set of parameters, possible forcing low data rate optimization on or off.
+    /// Low data rate optimization is determined automatically
+    /// based on `sf` and `bw` according to Semtech's datasheets for SX126x/SX127x
+    /// (enabled if symbol length is >= 16.38ms)
+    pub const fn new(sf: SpreadingFactor, bw: Bandwidth, cr: CodingRate) -> Self {
```
Comment (ilya-epifanov) [inline modulation/src/lib.rs:117]: new API

File: modulation/src/lib.rs [hunk 6]
```diff
@@ -72,3 +84,169 @@ pub enum CodingRate {
...
+        // according to SX127x 4.1.1.6 it's 16ms
+        // SX126x says it's 16.38ms
+        // probably it's 16.384ms which is SF11@125kHz
+        let ldro = t_sym_us >= 16_384;
+        Self { sf, bw, cr, ldro, t_sym_us }
+    }
+
+    /// Calculates time on air for a given payload and modulation parameters.
+    /// If `preamble` is None, the whole preamble including syncword is excluded from calculation.
+    pub const fn time_on_air_us(
```
Comment (ilya-epifanov) [inline modulation/src/lib.rs:128]: new API

**Pull request merged**